onnx/shape_inference/implementation.cc

```diff
@@ -439,69 +459,67 @@ class ShapeInferenceImplBase {
+        if (options.enable_data_propagation && schema && schema->has_data_propagation_function()) {
+          if (generated_shape_data_by_name == nullptr) {
+            fail_shape_inference(
+                "Container for generated shape data cannot be nullptr when enable_data_propagation option is set.");
+          }
+          DataPropagationContextImpl data_propagation_ctx(
+              n, value_types_by_name, input_data_by_name, *generated_shape_data_by_name);
+          schema->GetDataPropagationFunction()(data_propagation_ctx);
+        }
       }
     }
     ONNX_CATCH(const ONNX_NAMESPACE::InferenceError& ex) {
       ONNX_HANDLE_EXCEPTION([&]() {
-        // onnx does not support unsupported/experimental operators
-        // so it won't consider it as an error
-        if (!has_unsupported_op && !has_experimental_op) {
+        // Note: The following special handling is to accommodate custom-ops. Ideally, custom-ops
+        // should be registered with a schema in the schema registry, allowing inference to handle
+        // them. As things stand, this special handling is somewhat fragile and is not fully
+        // general either. Eg., a custom-op suppresses error-messages for subsequent nodes, but
+        // this does not work across graphs. If special handling is required, a user-option may
+        // be a better way to do it. The fragility comes from the fact that the types of the
+        // returned-values of the custom-op are unknown, and subsequent node-level inference
+        // may fail because of this.
+        if (!has_unsupported_op) {
           inference_errors.push_back(GetErrorWithNodeInfo(n, ex));
         }
       });
-      // Continue with inference for remaining nodes
-      return;
-    }
-
-    ONNX_TRY {
-      // check the type-equality for input and output
-      if (options.check_type && schema) {
-        schema->CheckInputOutputType(ctx);
-      }
-
-      for (int i = 0; i < n.output_size(); ++i) {
-        // skip type and shape propagation for missing optional outputs.
-        if (!n.output(i).empty())
-          updateType(n.output(i), ctx.getOutputType(i));
-      }
-
-      preprocess(n);
-
-      // If data propagation is enabled, propagate shape data if it exists.
-      if (options.enable_data_propagation && schema && schema->has_data_propagation_function()) {
-        if (generated_shape_data_by_name == nullptr) {
-          fail_shape_inference(
-              "Container for generated shape data cannot be nullptr when enable_data_propagation option is set.");
-        }
-        DataPropagationContextImpl data_propagation_ctx(
-            n, value_types_by_name, input_data_by_name, *generated_shape_data_by_name);
-        schema->GetDataPropagationFunction()(data_propagation_ctx);
-      }
     }
     ONNX_CATCH(const std::runtime_error& err) {
+      // TODO: Fix this. Unclear if this should be remapped to a shape inference error.
```

Comment:
Should we create an issue and link it from here for better tracking?

Reply:
It is related to the cleanup that this PR https://github.com/onnx/onnx/pull/5519 is doing. I can add a link to that PR